includes/Admin/Cross_Origin_Isolation.php

```diff
@@ -0,0 +1,292 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\Admin;
+
+use Google\Web_Stories\User_Preferences;
+use Google\Web_Stories\Story_Post_Type;
+use Google\Web_Stories_Dependencies\AmpProject\Dom\Document;
+use Google\Web_Stories\Experiments;
+use WP_Screen;
+
+/**
+ * Class Cross_Origin_Isolation
+ *
+ * @package Google\Web_Stories
+ */
+class Cross_Origin_Isolation {
+
+	/**
+	 * Experiments instance.
+	 *
+	 * @var Experiments Experiments instance.
+	 */
+	private $experiments;
+
+	/**
+	 * Cross_Origin_Isolation constructor.
+	 *
+	 * @param Experiments $experiments Experiments instance.
+	 */
+	public function __construct( Experiments $experiments ) {
+		$this->experiments = $experiments;
+	}
+	/**
+	 * Init
+	 *
+	 * @return void
+	 */
+	public function init() {
+		if ( ! $this->is_needed() || ! $this->is_edit_screen() ) {
+			return;
+		}
+
+		add_action( 'admin_footer-post.php', [ $this, 'admin_footer' ] );
+		add_action( 'admin_footer-post-new.php', [ $this, 'admin_footer' ] );
+		add_action( 'load-post.php', [ $this, 'admin_header' ] );
+		add_action( 'load-post-new.php', [ $this, 'admin_header' ] );
+		add_filter( 'style_loader_tag', [ $this, 'style_loader_tag' ], 10, 3 );
+		add_filter( 'script_loader_tag', [ $this, 'script_loader_tag' ], 10, 3 );
+		add_filter( 'get_avatar', [ $this, 'get_avatar' ], 10, 6 );
+	}
+
+	/**
+	 * Start output buffer and add headers.
+	 *
+	 * @since 1.6.0
+	 *
+	 * @return void
+	 */
+	public function admin_header() {
+		header( 'Cross-Origin-Opener-Policy: same-origin' );
+		header( 'Cross-Origin-Embedder-Policy: require-corp' );
+
+		ob_start();
+	}
+
+	/**
+	 * Do string replacement and output.
+	 *
+	 * @since 1.6.0
+	 *
+	 * @return void
+	 */
+	public function admin_footer() {
+		$html = (string) ob_get_clean();
+
+		$document = Document::fromHtml( $html );
+
+		if ( ! $document ) {
+			return;
+		}
+
+		$map = [
+			'link'   => 'href',
+			'img'    => 'src',
+			'iframe' => 'src',
+			'script' => 'src',
+		];
+
+		foreach ( $map as $tag => $attribute ) {
+			$tags = $document->getElementsByTagName( $tag );
+			foreach ( $tags as $node ) {
+				$value = $node->getAttribute( $attribute );
+				if ( ! $value ) {
+					continue;
+				}
+
+				$cross_origin = $node->getAttribute( 'crossorigin' );
+
+				if ( $cross_origin ) {
+					continue;
+				}
+
+				$html = $this->add_attribute( $html, $attribute, $value );
+			}
+		}
```

Comment:
This nested loop doesn't seem to do what's intended here, or at least does so very inefficiently. 

Expected:

1. Loop through every single link/img/iframe/script tag in the document
2. For N tags, check if the tag has the `crossorigin` attribute or add it to **this single tag** if needed.

What it does:

1. Loop through every single link/img/iframe/script tags in the document.
2. For N tags, call `$this->add_attribute()` on the **whole document**.

So if there are 100 tags, the whole document is updated 100 times in this loop, making this quite expensive.

To fix:

1. Use `$node->setAttribute()`
2. Echo `$document->saveHTML()` at the end

Reply:
> Use $node->setAttribute()
> Echo $document->saveHTML() at the end

Original I did this, but it changed the markup a lot, as php has processed the html and markup changed. An example was `<img />` tags were replaced with `<img >` tags. There were other changed to markup. I was worried here about messing with the doc to much. I know that some seemingly invalid html markup might be stripped for no good reason. 

Thoughts.

Reply:
Good point, and would be good to document this clearly in the inline documentation.

In that case, we should replace the `$html = $this->add_attribute( $html, $attribute, $value );` line with something along these lines:

```php
$original_element = $document->saveHTML( $node );
$modified_element = $this->add_attribute( $original_element, $attribute, $value );
$html = str_replace( $original_element, $modified_element, $html )
```

Reply:
So I tried this snippet and it didn't work. I added a unit test to prove that it didnt work here. 
https://github.com/google/web-stories-wp/pull/6875/commits/299b3a17230661ed6d465d99d85693bc650d7d9e

The issue again, is the self closing tags, that are converted. As the document element is modified and normalized, then replacement do not work correctly.